src/AbstractUri.php

```diff
@@ -362,30 +362,96 @@ protected function formatHost($host)
             return $host;
         }
 
-        static $pattern = '/^(?<name>[a-z0-9]([a-z0-9-]{0,61}[a-z0-9])?)(?:\.(?&name))*$/i';
-        if (preg_match($pattern, $host)) {
-            return strtolower($host);
+        if ('[' === $host[0] && ']' === substr($host, -1)) {
+            return $this->formatIpv6($host);
         }
 
-        if (!is_host($host)) {
-            throw UriException::createFromInvalidHost($host);
+        return $this->formatRegisteredName($host);
+    }
+
+    /**
+     * Validate and Format the IPv6 host containing or not a ZoneId
+     *
+     * @param string $host
+     *
+     * @throws UriException if the submitted host is not a valid IPv6
+     *
+     * @return string
+     */
+    private function formatIpv6(string $host)
+    {
+        $ipv6 = substr($host, 1, -1);
+        if (filter_var($ipv6, FILTER_VALIDATE_IP, FILTER_FLAG_IPV6)) {
+            return $host;
+        }
+
+        if (false === ($pos = strpos($ipv6, '%'))) {
+            throw new UriException(sprintf('the submitted host %s is an Invalid IPV6 hostname', $host));
+        }
+
+        static $gen_delims = '/[:\/?#\[\]@ ]/'; // Also includes space.
+        if (preg_match($gen_delims, rawurldecode(substr($ipv6, $pos)))) {
+            throw new UriException(sprintf('the submitted host %s is an Invalid IPV6 hostname', $host));
+        }
+
+        $ipv6 = substr($ipv6, 0, $pos);
+        if (!filter_var($ipv6, FILTER_VALIDATE_IP, FILTER_FLAG_IPV6)) {
+            throw new UriException(sprintf('the submitted host %s is an Invalid IPV6 hostname', $host));
         }
 
-        if ('[' === $host[0]) {
+        $reducer = function (string $carry, string $char): string {
+            return $carry.str_pad(decbin(ord($char)), 8, '0', STR_PAD_LEFT);
+        };
+
+        static $local_link_prefix = '1111111010';
+        $res = array_reduce(str_split(unpack('A16', inet_pton($ipv6))[1]), $reducer, '');
+        if (substr($res, 0, 10) === $local_link_prefix) {
             return $host;
         }
```

Comment:
This code looks highly complicated, could you add a comment what exactly it does?

Reply:
updated with comments

Reply:
How about simply https://3v4l.org/lMRh5?

```php
var_dump(substr(inet_pton("fe80::1") & "\xfe\x80", 0, 2) === "\xfe\x80");
var_dump(substr(inet_pton("fe3a::1") & "\xfe\x80", 0, 2) === "\xfe\x80");
```